### ISSUE TITLE
Stuck for days on Query with Large Backup set?

### ISSUE THREAD

**Issue by robbcrg** [duplicati/duplicati#2815]
- [X] I have searched open and closed issues for duplicates.

#1174 : Similar symptom (Stuck Completing Backup) but not the same details
#1797 : Similar symptom (Stuck Completing Backup) but not the same details

----------------------------------------

## Environment info
- **Duplicati version**: 2.0.2.9_canary_2017-10-01
- **Operating system**: CentOS 6.7 Final
- **Backend**: Google Cloud Storage
- System: 
-- Intel(R) Xeon(R) CPU E5-2640 v3 @ 2.60GHz
-- 32 Cores
-- 32 GB RAM with at least 11 GB available
-- Disk: SSD Raid  : 618 GB Freee
-- Duplicati is running on the IRON backing up a Virtuozza (OpenVZ) environment

## Description
Stuck for 3 days at Completing Backup. The last log entry shows the following query:

```
SELECT "CalcLen", "Length", "A"."BlocksetID", "File"."Path" FROM (
 SELECT
 "A"."ID" AS "BlocksetID",
 IFNULL("B"."CalcLen", 0) AS "CalcLen",
 "A"."Length"
 FROM
 "Blockset" A
 LEFT OUTER JOIN
 (
 SELECT
 "BlocksetEntry"."BlocksetID",
 SUM("Block"."Size") AS "CalcLen"
 FROM
 "BlocksetEntry"
 LEFT OUTER JOIN
 "Block"
 ON
 "Block"."ID" = "BlocksetEntry"."BlockID"
 GROUP BY "BlocksetEntry"."BlocksetID"
 ) B
 ON
 "A"."ID" = "B"."BlocksetID"

 ) A, "File" WHERE "A"."BlocksetID" = "File"."BlocksetID" AND "A"."CalcLen" != "A"."Length";

```

An explain can be seen at:
-  https://pastebin.com/cJeiCk0F

That SQL looks like it would be rather slow for a large database. My database is:

- DB Size: 5,881,373,696
- Files: 6,380,017
- Block: 10,361,766
- BlockSetID: 9,245,972
- BlockSetEntry: 10,389,116

Backup set on GCS is:
- 55,567,242,393 Bytes

- DB was last modified:
-- Oct 11  08:52 EDT 2017
- Current Server Time
-- Sat Oct 14 02:15:33 EDT 2017

Mono is shown running at 99% mostly on one CPU of a 32 core system.

```
  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND
40000 root      20   0 2131m 133m 2296 S 100.2  0.4   4388:27 mono
```

## Steps to reproduce
1. Backup a large set of files with a lot of duplicates
2. Watch it get stuck in the query

- **Actual result**:
Stuck "Completing Backup"

- **Expected result**:
Finishes

**Comment by robbcrg** [comment]
Just a  follow up: It seems Duplicati will never finish this -- even after days.

**Comment by CaptainKernel** [comment]
Hello, 

Still happening in 2.0.2.1_beta_2017-08-01, same issue with large OpenVZ container, stuck at "Completing Backup". Same OS, similar system specs.

Backing up /vz/private/$CTID from the host, running the task with:
--snapshot-policy=Auto
--no-backend-verification=true
--backup-test-samples=1
--blocksize=200KB
--overwrite=true
--skip-file-hash-checks=true
--symlink-policy=Store
--check-filetime-only=true
--disable-filepath-cache=true

**Comment by cjbnc** [comment]
I also ran into this problem. In my dataset, it takes about 85 minutes to complete that query, which is found in the code under Duplicati/Library/Main/Database/LocalDatabase.cs in the VerifyConsistency routine.
```
VerifyConsistency took 01:25:34.060

# some sizes from my dataset and the tables in that query:
select count(*) from Blockset;
210324
select count(*) from BlocksetEntry;
363132
select count(*) from Block;
368152
```

I played with the query in my sqlite3 database directly and confirmed the slowdown is due to one of the nested LEFT JOINs which is creating a 210000 row result with no index on the ID column used in the join.  I came up with this possible solution using a temporary table that is properly indexed. This code returns in a couple seconds, not 85 minutes:

```
CREATE TEMPORARY TABLE subq2 (
  "BlocksetID" INTEGER PRIMARY KEY,
  "CalcLen" INTEGER NOT NULL
);

INSERT INTO subq2
SELECT "BlocksetEntry"."BlocksetID" as "BlocksetID",
    SUM("Block"."Size") AS "CalcLen"
  FROM "BlocksetEntry"
  LEFT OUTER JOIN "Block" ON "Block"."ID" = "BlocksetEntry"."BlockID"
  GROUP BY "BlocksetEntry"."BlocksetID";

SELECT "CalcLen", "Length", "A"."BlocksetID", "File"."Path"
  FROM ( 
    SELECT "A"."ID" AS "BlocksetID", IFNULL("B"."CalcLen", 0) AS "CalcLen",
       "A"."Length" FROM "Blockset" A 
    LEFT OUTER JOIN "subq2" B ON "A"."ID" = "B"."BlocksetID"
  ) A, "File"
  WHERE "A"."BlocksetID" = "File"."BlocksetID"
  AND "A"."CalcLen" != "A"."Length";
```

Hope this helps.

**Comment by cjbnc** [comment]
I noticed @robbcrg was referring to a Centos6.7 server. My test above was on a RHEL6 workstation. I ran some further testing on a Fedora 23 box instead, and did not see the same timing problems in SQLite. In fact, I was able to copy the RHEL6 data file over to the newer FC23 machine and the query completed in no time.

It looks to me like having a newer version of SQLite on the server makes all the difference:
```
RHEL6: sqlite-3.6.20-1.el6_7.2.x86_64
Fedora 23: sqlite-3.11.0-3.fc23.x86_64
Current SQLite from source is 3.22.0
```

I was working on a patch to implement the temp table described above. It's available in my fork, if you want it: [cjbnc/duplicati](https://github.com/cjbnc/duplicati/tree/sql_improvements). IMHO this is unnecessary code if the underlying SQLite code can be updated.

**Comment by cjbnc** [comment]
Further confirmation on my RHEL6 machine. Visit https://sqlite.org/download.html and grab the precompiled 3.22 for linux binary. Run the same query on the same dataset. Returns in 0.753s.

**Comment by cjbnc** [comment]
My RHEL6 backups now finish in 3 minutes, not 5 hours. 
In summary:
RHEL/Centos 6 installs an old version (3.6.20) of SQLite that does not appear to correctly optimize queries containing complex joins. Looking though the SQLite history, I think the [NGQP](https://sqlite.org/queryplanner-ng.html) that they added with 3.8.0 may be the missing piece. Installing SQLite 3.8.8 from a 3rd-party RPM source solves the slow query problem on my RHEL6 machine.

If you are running RHEL/Centos6 and you are willing to include 3rd party RPMs on your server, here's what I did to fix my workstation:
```
# as root. I needed all three of these packages to replace existing installs
cd /tmp
wget ftp://ftp.pbone.net/mirror/ftp5.gwdg.de/pub/opensuse/repositories/Application:/Geo/RHEL_6/x86_64/sqlite-3.8.8.1-142.1.x86_64.rpm
wget ftp://ftp.pbone.net/mirror/ftp5.gwdg.de/pub/opensuse/repositories/Application:/Geo/RHEL_6/x86_64/sqlite-devel-3.8.8.1-142.1.x86_64.rpm
wget ftp://ftp.pbone.net/mirror/ftp5.gwdg.de/pub/opensuse/repositories/Application:/Geo/RHEL_6/i686/sqlite-3.8.8.1-142.1.i686.rpm
yum update sqlite*
# should upgrade to the files found in the current directory
# be sure to restart duplicate-server after the update
```

If you are not willing to replace the official RPMs with a newer, 3rd party version, I'd try running Duplicati in a docker container next. It's a path I've had to follow a few times when dealing with servers that have older, version-locked software like this. 

Another idea would be to manually build and install SQLite > 3.8.0 in a parallel root and then play with LD_LIBRARY_PATH or similar when starting Duplicati.

**Comment by cjbnc** [comment]
One more data point: Centos7 installs SQLite version 3.7.17 2013-05-20. Testing with my dataset on the sqlite3 command line finished in 1.8 seconds, which is quite reasonable. Looks like the optimization we need is already in 3.7.17, not 3.8.0 as I had guessed above.

**Comment by robbcrg** [comment]
Thanks for the research cjbnc - I also tested breaking the queries up and like you found it dramatically faster. But I decided to pursue you SQLite upgrade suggestion before offering a code update and found the solution worked quite well. In my case I compiled Sqlite 3.22 (latest) on a CentOS 6.7 machine and copied the needed files to my Duplicati working area, added the LD_LIBRARY_PATH (don;t forget export) to my startup script and found it worked wonderfully. This is a server so updating the core SQLite was not something I wanted to risk and a docker session was overkill.
